README
NAME
    Alt::Crypt::RSA::BigInt - Rewrite of Crypt::RSA

DESCRIPTION

    This is a rework of Crypt::RSA to remove Math::Pari and instead base
    all operations on Math::BigInt.

STATUS

    This is intended to be a plug-in replacement for Crypt::RSA, with no
    user-visible changes to existing features.

    Math::Pari is completely removed.  This includes the two modules:

       - Crypt::Primes   =>   Math::Prime::Util
       - Crypt::Random   =>   Bytes::Random::Secure

    All operations are now performed using Math::BigInt, and prefer the
    GMP and Pari backends.

    All of the existing bug reports have been addressed, as well as adding
    a few new features such as 

PERFORMANCE

    Performance using GMP is 3-10 times faster than the Crypt::RSA 1.99.

    Using Math::BigInt::Pari, it is about half the speed at signing, and
    on par when verifying.

    If neither GMP nor Pari are available, performance is very slow, from
    10x to 200x slower.  However this is an environment where the original
    code could not run.

    Time to run test suite on my machine:
         4 seconds   with GMP and Math::Prime::Util::GMP
        10 seconds   with GMP
        17 seconds   with Pari (no GMP)
        91 seconds   with Calc (no GMP and no Pari)
    The test suite uses smaller sizes in some tests for non-GMP, non-Pari.

AUTHOR

    Vipul Ved Prakash wrote the original Crypt::RSA.

    Dana Jacobsen did the changes to Math::BigInt.

COPYRIGHT AND LICENSE

    Copyright (c) 2001 by Vipul Ved Prakash, 2012 by Dana Jacobsen.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.


Changes
Revision history for Perl module Alt::Crypt::RSA::BigInt

0.07  ?

  - Return true value from Alt module.  RT 105786.


0.06  2017-04-26

  - Typo fixes from dsteinbrunner.

  - Use new Math::Prime::Util with a CSPRNG, remove Bytes::Random::Secure.

  - Add check in key gen for p/q distance (FIPS, ANSI), for size over 256.

  - Use BPSW + Frobenius + 3 random-base M-R instead of BPSW + 46 fixed M-R.
    No practical difference, but a little faster.


0.05  2013-02-22

  - Fixed some perlcritic issues -- mainly returning undef and changing to the
    three argument open using a lexical file handle.

  - Doh -- Alt's package name was Bigint not BigInt.

  - Speed up testing for Calc backend (only do 384-bits in test 11 if not
    using GMP or Pari backends).

  - Switch to Bytes::Random::Secure's OO interface, and allow plugging in a
    different irand function to the generator.

  - Use the new generator interface in all tests.  The suite went from
    3328 bits of blocking randomness to 1536.


0.04  2013-01-14

  - Change from random_maurer_prime to random_nbit_prime.  This is much faster
    and yields more possible primes (Maurer's algorithm only outputs about 10%
    of possible primes in the range).  The downside is that the primes are not
    proven prime, however they will have passed a strong BPSW test and a set of
    Miller-Rabin tests for the first 46 prime bases.  The BPSW test alone is
    really enough for these sizes (~600 digits or smaller).

  - Ask for Math::BigInt 1.78 to ensure we have 'try'.
    Thanks to brian carlson.

  - Reduce key sizes in 06 and 11 when not using GMP or Pari.  Some build
    systems were timing out (e.g. ActiveState).


0.03  2012-12-30

  - Add tests for SHA384 and SHA512.

  - Add optional support for RIPEMD160.

  - Add more data to meta file, including git repository.

  - All tests converted to Test::More.

  - Rewrite Key/Private/SSH.pm, and add tests.  I don't think it ever worked.
    Now supports IDEA, DES, DES3, Blowfish, Twofish, CAST5, Rijndael, RC6, and
    Camelia (only the first 4 are guaranteed -- the others depend on having a
    module installed).  Multiple alternative modules are allowed for some.
    This fixes:

     RT 69376  Bug in Crypt::RSA::Key::Private::SSH serialize function

  - FindBin is no longer used.  You'll need to do:
       perl -Iblib/lib -Iblib/arch ...
    to get the version in this directory.  'make test' does this for us.
    use 'prove -l -b' to prove using the local version.

    I think FindBin is great for example files and such, but I don't think
    it is appropriate in the .pm files or in tests.  If we do use it, it has
    to be added to the dependency list.


0.02  2012-12-28

  - Fix a decrypt issue setting keys

  - Add SHA224 support

  - Added strict and warnings to all .pm files.

  - Converted RSA.pm to UTF-8 and added POD directive

  - Make pretty maurer prime generation output with Verbosity.


0.01  2012-12-28

  - Add test for h2osp, pss sign and verify with salt.

  - Faster versions of some of the helper functions.

  - Switch from Digest::SHA1 to Digest::SHA, and add SHA256.

  - The following Crypt::RSA 1.99 defects have been fixed or made irrelevant
    due to no longer using Math::Pari:

     RT 52689  consider moving Crypt::RSA to Math::BigInt?
     RT 76655  Version 1.99's META.yml says version is 1.97
     RT 63007  Move to later version of Math-PARI
     RT  4877  Crypt::RSA test 11 fails with a segmentation fault
     RT 61392  Crypt::RSA::Key::Private::check not checking that q is an integer
     RT 64883  Make errstr from check after write more intuitive
     RT 45533  Crypt::RSA bug in mgf1 function

    These fixable possible defects remain, but should be resoved soon.

     RT 69376  Bug in Crypt::RSA::Key::Private::SSH serialize function

    as well as these that aren't clear:

     RT 50356  Crypt-RSA-1.99
     RT 59923  changing the passphrase on key

  - Makefile.PL will look for GMP and Pari, and install the Math::BigInt::GMP
    or Math::BigInt::Pari modules if possible.  It complains otherwise, but
    everything should still work -- it just runs really slow.

  - Completely removed Pari.

  - Replaced modules:

       Crypt::Random  =>  Bytes::Random::Secure
       Crypt::Primes  =>  Math::Prime::Util

  - Started with Vipul's Crypt::RSA.


Makefile.PL
use ExtUtils::MakeMaker;
use lib 'inc'; # load our bundled version of Devel::CheckLib
use Devel::CheckLib;

my %bigint_preq;

if ( check_lib(lib => 'gmp', header => 'gmp.h') ) {
  # GMP found.  Life is *good*.
  %bigint_preq = (
                      'Math::BigInt::GMP'      => 0,
                      'Math::Prime::Util::GMP' => 0,
                 );
} elsif ( check_lib(lib => 'pari', header => 'pari/pari.h') ) {
  # Pari found.  Slower than GMP in this application, but still adequate.
  %bigint_preq = (
                      'Math::BigInt::Pari'     => 0,
                 );
} else {
  # Yech, no big int libraries found.
  print <<EO_BIGINT;
 

 **********************************************************
 *
 * Building without GMP or Pari support.
 *
 * Operations will be very slow.
 *
 * Strongly consider installing GMP and Math::BigInt::GMP.
 *
 **********************************************************
EO_BIGINT
}

WriteMakefile1(
    NAME         => 'Alt::Crypt::RSA::BigInt',
    ABSTRACT     => 'RSA public-key cryptosystem, using Math::BigInt',
    VERSION_FROM => 'lib/Alt/Crypt/RSA/BigInt.pm',
    LICENSE      => 'perl',
    AUTHOR       => 'Dana A Jacobsen <dana@acm.org>',

    BUILD_REQUIRES=>{
                      'Test::More'       => '0.45',
                      'Benchmark'        => 0,
                      'MIME::Base64'     => 0,
                    },
    PREREQ_PM    => {
                      'Math::BigInt'           => '1.78', # try => backend
                      'Math::Prime::Util'      => '0.64', # built in csprng
                      'Carp'                   => 0,
                      'Exporter'               => 0,
                      'Crypt::CBC'             => '2.17',
                      'Crypt::Blowfish'        => 0,
                      'Data::Dumper'           => 0,
                      'Tie::EncryptedHash'     => 0,
                      'Convert::ASCII::Armour' => 0,
                      'Sort::Versions'         => 0,
                      'Digest::SHA'            => 0,
                      'Digest::MD5'            => 0,
                      'Digest::MD2'            => 0,
                      'Class::Loader'          => 0,
                      'Data::Buffer'           => 0,
                      %bigint_preq,
                    },

    # This should be worked on.  For now, no 5.6.x.
    MIN_PERL_VERSION => 5.008,
    META_MERGE   => {
                      no_index   => { directory => [qw(lib/Crypt)] },
                      resources  => {
                         homepage   => 'https://github.com/danaj/Alt-Crypt-RSA-BigInt/tree',
                         repository => 'git://github.com/danaj/Alt-Crypt-RSA-BigInt.git',
                       },
                      recommends => {
                        'Crypt::RIPEMD160' => 0.05,
                      },
                    },
);

sub WriteMakefile1 {   # Cribbed from eumm-upgrade by Alexandr Ciornii
  my %params = @_;
  my $eumm_version = $ExtUtils::MakeMaker::VERSION;
  $eumm_version = eval $eumm_version;

  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
      #EUMM 6.5502 has problems with BUILD_REQUIRES
      $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
      delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;
  delete $params{AUTHOR} if $] < 5.005;
  delete $params{ABSTRACT_FROM} if $] < 5.005;
  delete $params{BINARY_LOCATION} if $] < 5.005;

  WriteMakefile(%params);
}
